### ISSUE TITLE
Duplicate tasks created for workflow

### ISSUE THREAD

**Issue by raghavendra-bhat** [Netflix/conductor#2098]
We are observing a strange behaviour during our load test which we didn't find during regular functional testing. The workflows are getting terminated as a task (T5) could not be processed as it is unable to get the correct input. The task's input is dependent on a previous task's (T2) output. UI marks T2 task as red indicating it failed but a look at the task details tab in the UI shows that task was completed. However a deeper look in to the Cassandra table "workflows" shows that there were actually two tasks that got created for the same task type (T2). Please see the attached screenshots

<img width="1176" alt="Screenshot 2021-02-19 at 5 03 05 PM" src="https://user-images.githubusercontent.com/15682929/108529259-4e8a2980-72fa-11eb-8bf3-5b32c7ad7638.png">
<img width="1529" alt="Screenshot 2021-02-19 at 5 03 12 PM" src="https://user-images.githubusercontent.com/15682929/108529286-52b64700-72fa-11eb-8348-f806fffa9a3a.png">

We are using Cassandra persistence module and trying out the QueueDao implementation using Hazelcast. 
**NOTE**: These tests are running without any lock with disabled workflow sweeper service. 

Below is the workflows table (along with start time for the duplicate tasks)

| workflow_id                          | shard_id | entity   | task_id                              |start time| sch time| total_partitions | total_tasks|
|---------------------------|:---------|:----------|:--------------------------------------|:----:|:----:|:-----------------|:-------------|
| 93e14b7a-6ac3-4224-bd6d-25001234b15d |        1 |     task | 19eb12ce-cf30-41ac-bc89-2e769e9e3f08 || |                1 |           6|
| 93e14b7a-6ac3-4224-bd6d-25001234b15d |        1 |     task | 304c5813-36b2-465d-b866-55c2753a964e (**T5**)|| |                1 |           6|
| 93e14b7a-6ac3-4224-bd6d-25001234b15d |        1 |     task | c2e9902a-52c5-4a27-9f49-0333d42378c5 || |                1 |           6|
| 93e14b7a-6ac3-4224-bd6d-25001234b15d |        1 |     task | c5e5ae73-efe1-494e-9ff5-fdb6f0271d96 || |                1 |           6|
| 93e14b7a-6ac3-4224-bd6d-25001234b15d |        1 |     task | d2320dc7-57da-4b7b-9775-22a0bba456f5 || |                1 |           6
| 93e14b7a-6ac3-4224-bd6d-25001234b15d |        1 |     task | **d5d929ed-129d-4e4a-8fad-04180b83c77e** (**T2**) |1613657949180| 1613657949192|                1 |           6|
| 93e14b7a-6ac3-4224-bd6d-25001234b15d |        1 |     task | **e4dc89ee-c9dd-4a7a-8468-4cb7c6ab8cde**  (**T2**) |1613657949175| 1613657949190|                1 |           6|
| 93e14b7a-6ac3-4224-bd6d-25001234b15d |        1 | workflow |                                      || |                1 |           6|

The **bold** highlighted records are of the duplicate tasks of the same task type/task ref. Notice that the start and schedule time of these two task instances are very close to each other as well. As these tasks are not idempotent, running the task a second time results in an error. One of the task instance is executed with COMPLETED status while the other gets CANCELED and it doesn't the get input either. The workflow moves ahead after the task instance got completed and reaches the task T5 at which point the duplicate task is canceled. The T5 task in the workflow is dependent on the T2's output and fails as it doesn't have the necessary input. It appears that it is getting the duplicate task's output instead of the successful ones and hence it fails.

I have the following concerns:
1. Why would Conductor create duplicate task instances? 
2. How can I proceed to resolve the issue so that the duplicate tasks are not getting created in the first place?
3. Does not having a lock play any role in such anomaly? 

Thanks in advance on any pointers.

**Comment by aravindanr** [comment]
Please always include workflow and task definitions when posting a question. Without those, we can only speculate. In the data you have shared, the scheduled is after the start time, which does not happen. Is it possible that the columns got interchanged when posting the question?

> 1. Why would Conductor create duplicate task instances?

Conductor creates a task instance for each attempt of the task. If a task has 3 retries, each attempt gets its own task instance. From the UI screenshots, that doesn't seem to be the case here.

> 3. Does not having a lock play any role in such anomaly?

It is quite possible that this is the case here. You have noted that the sweeper is disabled. But if you have Conductor running in multiple instances, the "decide" process _could_ step on each other. 

> One of the task instance is executed with COMPLETED status while the other gets CANCELED and it doesn't the get input either. The workflow moves ahead after the task instance got completed and reaches the task T5 at which point the duplicate task is canceled. 

If one of them is CANCELED, how does the UI happen to pick the entry in COMPLETED? Does the state change on refresh of the page?

It's hard for us to replicate your exact setup and reproduce the issue. We do recommend that lock is enabled as it guarantees only one thread (across all instances) is operating on a workflow object. The flag to disable lock exists mainly for convenience and testing. 
We can say with certainty that duplicate tasks do not get created with Cassandra persistence and lock enabled.

**Comment by raghavendra-bhat** [comment]
Hi @aravindanr, Sorry for delay in getting you the details. We have been trying to replicate the issue again, this time by introducing Zookeeper lock and having the workflow sweeper service enabled. Even after introducing locks, I'm seeing that the duplicate tasks get created. They get created at different tasks in different failed workflow executions. I have noticed that if the workers are able to pull the duplicate tasks then I see that they get completed else the duplicate task get CANCELED. Not sure who is doing the cancellation. Please find attached screenshot of the workflow that shows this issue.  

<img width="1685" alt="Screenshot 2021-02-26 at 7 19 10 PM" src="https://user-images.githubusercontent.com/15682929/109308121-9a435280-7867-11eb-9506-4f655bec814c.png">
<img width="1760" alt="Screenshot 2021-02-26 at 7 19 17 PM" src="https://user-images.githubusercontent.com/15682929/109308717-66b4f800-7868-11eb-841a-0f493480fb93.png">


> If one of them is CANCELED, how does the UI happen to pick the entry in COMPLETED? Does the state change on refresh of the page?

The UI picks up the CANCELED task instead of the completed ones. Doesn't change after refresh.

> In the data you have shared, the scheduled is after the start time, which does not happen. Is it possible that the columns got interchanged when posting the question?

I have a better view of of the  Cassandra "workflows" table view for the tasks in this workflow instance attached below. The pattern that we have seen is that scheduled time difference between the duplicate tasks is in the order of few milliseconds.

<img width="1020" alt="Screenshot 2021-02-26 at 7 29 40 PM" src="https://user-images.githubusercontent.com/15682929/109309220-feb2e180-7868-11eb-9e6f-4006336c60a3.png">


This happens when the system is under load. I haven't observed this issue when running a single user test.
I am also attaching the workflow and task meta info with this.

[meta-defs.zip](https://github.com/Netflix/conductor/files/6049997/meta-defs.zip)

**Comment by raghavendra-bhat** [comment]
@aravindanr I think we were able to figure out the cause for the duplicate tasks issue. Zookeeper cluster misconfiguration in our perf environment resulted in this issue. Zookeeper nodes were giving locks to the same workflow for two different transactions. Workflow sweeper was also getting a lock on the same workflow that was being processed by the start workflow/update task transaction. As a result duplicate tasks were created as both of them call the same decide method. After we have configured the Zookeeper to form a correct ensemble cluster we no longer see this issue. Sorry for the trouble caused because of the bad configuration.
